src/tools_for_rules.jl

```diff
@@ -0,0 +1,477 @@
+macro mooncake_overlay(method_expr)
+    def = splitdef(method_expr)
+    def[:name] = Expr(:overlay, :(Mooncake.mooncake_method_table), def[:name])
+    return esc(combinedef(def))
+end
+
+#
+# Functionality supporting @zero_adjoint
+#
+
+"""
+    zero_adjoint(f::CoDual, x::Vararg{CoDual, N}) where {N}
+
+Utility functionality for constructing `rrule!!`s for functions which produce adjoints which
+always return zero.
+
+NOTE: you should only make use of this function if you cannot make use of the
+[`@zero_adjoint`](@ref) macro.
+
+You make use of this functionality by writing a method of `Mooncake.rrule!!`, and
+passing all of its arguments (including the function itself) to this function. For example:
+```jldoctest
+julia> import Mooncake: zero_adjoint, DefaultCtx, zero_fcodual, rrule!!, is_primitive, CoDual
+
+julia> foo(x::Vararg{Int}) = 5
+foo (generic function with 1 method)
+
+julia> is_primitive(::Type{DefaultCtx}, ::Type{<:Tuple{typeof(foo), Vararg{Int}}}) = true;
+
+julia> rrule!!(f::CoDual{typeof(foo)}, x::Vararg{CoDual{Int}}) = zero_adjoint(f, x...);
+
+julia> rrule!!(zero_fcodual(foo), zero_fcodual(3), zero_fcodual(2))[2](NoRData())
+(NoRData(), NoRData(), NoRData())
+```
+
+WARNING: this is only correct if the output of `primal(f)(map(primal, x)...)` does not alias
+anything in `f` or `x`. This is always the case if the result is a bits type, but more care
+may be required if it is not.
+```
+"""
+@inline function zero_adjoint(f::CoDual, x::Vararg{CoDual, N}) where {N}
+    return zero_fcodual(primal(f)(map(primal, x)...)), NoPullback(f, x...)
+end
+
+"""
+    @zero_adjoint ctx sig
+
+Defines `is_primitive(context_type, sig) = true`, and defines a method of
+`Mooncake.rrule!!` which returns zero for all inputs.
+Users of ChainRules.jl should be familiar with this functionality -- it is morally the same
+as `ChainRulesCore.@non_differentiable`.
+
+For example:
+```jldoctest
+julia> using Mooncake: @zero_adjoint, DefaultCtx, zero_fcodual, rrule!!, is_primitive
+
+julia> foo(x) = 5
+foo (generic function with 1 method)
+
+julia> @zero_adjoint DefaultCtx Tuple{typeof(foo), Any}
+
+julia> is_primitive(DefaultCtx, Tuple{typeof(foo), Any})
+true
+
+julia> rrule!!(zero_fcodual(foo), zero_fcodual(3.0))[2](NoRData())
+(NoRData(), 0.0)
+```
+
+Limited support for `Vararg`s is also available. For example
+```jldoctest
+julia> using Mooncake: @zero_adjoint, DefaultCtx, zero_fcodual, rrule!!, is_primitive
+
+julia> foo_varargs(x...) = 5
+foo_varargs (generic function with 1 method)
+
+julia> @zero_adjoint DefaultCtx Tuple{typeof(foo_varargs), Vararg}
+
+julia> is_primitive(DefaultCtx, Tuple{typeof(foo_varargs), Any, Float64, Int})
+true
+
+julia> rrule!!(zero_fcodual(foo_varargs), zero_fcodual(3.0), zero_fcodual(5))[2](NoRData())
+(NoRData(), 0.0, NoRData())
+```
+Be aware that it is not currently possible to specify any of the type parameters of the
+`Vararg`. For example, the signature `Tuple{typeof(foo), Vararg{Float64, 5}}` will not work
```

Comment:
Will it error in some understandable way?

Reply:
Hmmmm probably not. Alas, I'm not really sure how to ensure that it does either -- it's just a tricky one to design good code for.

Reply:
@mhauru if you have any thoughts on how I might catch things like `Tuple{typeof(foo), Vararg{T}} where {T}` and `Tuple{typeof(foo), Vararg{T} where {T}}`, and `Tuple{typeof(foo), Vararg{<:Real}}` etc, I'd love to hear it. 

Users of this macro will certainly get some kind of error though, so a mistake can't silently propagate through to users or anything nasty like that.

Reply:
I doubt I'll have much useful to add, don't let this hold you back from merging. Good to know that it'll at least error in some way.

Reply:
Cool. Thanks!